src/tribler-gui/tribler_gui/error_handler.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import os
+import traceback
+
+from tribler_common.sentry_reporter.sentry_reporter import SentryReporter
+
+from tribler_gui.dialogs.feedbackdialog import FeedbackDialog
+from tribler_gui.event_request_manager import CoreConnectTimeoutError
+
+
+class ErrorHandler:
+    def __init__(self, tribler_windows):
+        logger_name = self.__class__.__name__
+        self._logger = logging.getLogger(logger_name)
+        SentryReporter.ignore_logger(logger_name)
+
+        self.tribler_windows = tribler_windows
+
+        self._handled_exceptions = set()
+        self._tribler_stopped = False
+
+    def gui_error(self, *exc_info):
+        if self._tribler_stopped:
+            return
+
+        info_type, info_error, tb = exc_info
+        if info_type in self._handled_exceptions:
+            return
+        self._handled_exceptions.add(info_type)
+
+        text = "".join(traceback.format_exception(info_type, info_error, tb))
+
+        if info_type is CoreConnectTimeoutError:
+            text = text + self.tribler_windows.core_manager.core_traceback
+            self._stop_tribler(text)
+
+        self._logger.error(text)
+
+        FeedbackDialog(
+            parent=self.tribler_windows,
+            exception_text=text,
+            tribler_version=self.tribler_windows.tribler_version,
+            start_time=self.tribler_windows.start_time,
+            sentry_event=SentryReporter.event_from_exception(info_error),
+            error_reporting_requires_user_consent=True,
+            stop_application_on_close=self._tribler_stopped
+        ).show()
+
+    def core_error(self, text, core_event):
+        if self._tribler_stopped:
+            return
+
+        self._logger.error(text)
+
+        self._stop_tribler(text)
+
+        FeedbackDialog(
+            parent=self.tribler_windows,
+            exception_text=text,
+            tribler_version=self.tribler_windows.tribler_version,
+            start_time=self.tribler_windows.start_time,
+            sentry_event=core_event['sentry_event'],
+            error_reporting_requires_user_consent=core_event['error_reporting_requires_user_consent'],
+            stop_application_on_close=self._tribler_stopped
+        ).show()
+
+    def _stop_tribler(self, text):
+        if self._tribler_stopped:
+            return
+
+        self._tribler_stopped = True
+
+        self.tribler_windows.tribler_crashed.emit(text)
+        self.tribler_windows.delete_tray_icon()
+
+        # Stop the download loop
+        self.tribler_windows.downloads_page.stop_loading_downloads()
+
+        # Add info about whether we are stopping Tribler or not
+        os.environ['TRIBLER_SHUTTING_DOWN'] = str(self.tribler_windows.core_manager.shutting_down)
```

Comment:
I understand that this PR just move this code from some other place, but it probably should be fixed anyway.
When Tribler code checks the value of this environment variable, it expects to get "TRUE" or something else:
```python
@web.middleware
async def error_middleware(request, handler):
    try:
        if os.environ.get('TRIBLER_SHUTTING_DOWN', "FALSE") == "TRUE":
            raise Exception('Tribler is shutting down')
```
The values of environment variables are case-sensitive, at least on non-Windows systems. So the check above should not work when the actual value is `"True"` instead of `"TRUE"`

So I suspect you need to add `upper`:
```python
os.environ['TRIBLER_SHUTTING_DOWN'] = str(self.tribler_windows.core_manager.shutting_down).upper()
```

Reply:
Actually, it is not obvious to me, why we use this `TRIBLER_SHUTTING_DOWN` environment variable at all, instead of just setting some flag in memory. It may be useful when two processes are communicating using an environment variable value, but in our case, GUI does not read this environment variable value in its code.

Reply:
@devos50 seems to be an author of this line :)

Reply:
@kozlovsky changed to
```
os.environ['TRIBLER_SHUTTING_DOWN'] = str(self.tribler_windows.core_manager.shutting_down).upper()
```